created new section, how to set up development environment (#537)
kiali/kiali.io: content/en/docs/Contribution-guidelines/development-environment.md

  - [go](https://go.dev/)
  - [make](https://www.gnu.org/software/make/)
  - [npm](https://www.npmjs.com/)
  - [yarn](https://yarnpkg.com/)
- Istio and required services are running in Minikube or open shift. To install it following the above schema, it is possible to use the following scripts:
  - `hack/istio/install-istio-via-istioctl.sh`: Installs the latest Istio release into *istio-system* namespace along with the Prometheus, Grafana, and Jaeger addons.
  - `hack/istio/install-bookinfo-demo.sh`: Installs the Bookinfo demo that is found in the Istio release that was installed via the *hack/istio/install-istio-via-istioctl.sh* hack script.
    - Pass in `-tg` to also install a traffic generator that will send messages periodically into the Bookinfo demo.
    - If using Minikube and the `-tg` option, make sure you pass in the Minikube profile name via `-mp` if the profile name is not `minikube`.

### Port forward

Before the setup, we will need to do a port-forward of the services that kiali is using.

We can use the `hack/run-kiali.sh` script for this purpose. To run it following the above schema:

```shell

> ./run-kiali.sh -pg 13000:3000 -pp 19090:9090 -app 8080 -es false -iu http://127.0.0.1:15014

Reviewer: just a comment - reading this, it makes it seem as though the user NEEDS those options. I rarely pass any options and it works for me. Perhaps explain what those options do (and mention they are really optional).

Reviewer: Or just say, "It can work without any options. Pass `--help` to see the options it takes"

Author: Changes added.

---
title: "Development Environment"
description: "How to set up a development environment"
---

### Introduction

In this section it is explained how to set up a development environment:

> - As described in [Arquitecture](/docs/architecture/architecture), we would need to have the Kiali dependencies running in an Openshift or Kubernetes

Reviewer: Fix misspelling, and use "OpenShift" as one word, with uppercase O and S.

```suggestion
- As described in [Architecture](/docs/architecture/architecture), we would need to have the Kiali dependencies running in an OpenShift or Kubernetes
```

- We will use a port forward to access those services outside the cluster.
- We will have the project source running locally. In this case we will set up an IDE.
- Bookinfo application example will also be running on our cluster.

![development_environment](/images/documentation/contribution/development_environment.png)

> ### Prerrequisites 

Reviewer: fix typo:

```suggestion
### Prerequisites 
```

- Development tools are installed:

>   - [oc](https://docs.openshift.com/container-platform/4.7/cli_reference/openshift_cli/getting-started-cli.html) or [kubectl](https://kubernetes.io/es/docs/tasks/tools/)

Reviewer: We can update this to refer to the latest version of the doc:

```suggestion
  - [oc](https://docs.openshift.com/container-platform/4.10/cli_reference/openshift_cli/getting-started-cli.html) or [kubectl](https://kubernetes.io/es/docs/tasks/tools/)
```

  - [go](https://go.dev/)
  - [make](https://www.gnu.org/software/make/)
  - [npm](https://www.npmjs.com/)
  - [yarn](https://yarnpkg.com/)

> - Istio and required services are running in Minikube or open shift. To install it following the above schema, it is possible to use the following scripts:

Reviewer: Use the "official" way to mention OpenShift:

```suggestion
- Istio and required services are running in Minikube or OpenShift. To install it following the above schema, it is possible to use the following scripts:
```

  - `hack/istio/install-istio-via-istioctl.sh`: Installs the latest Istio release into *istio-system* namespace along with the Prometheus, Grafana, and Jaeger addons.
  - `hack/istio/install-bookinfo-demo.sh`: Installs the Bookinfo demo that is found in the Istio release that was installed via the *hack/istio/install-istio-via-istioctl.sh* hack script.
    - Pass in `-tg` to also install a traffic generator that will send messages periodically into the Bookinfo demo.
    - If using Minikube and the `-tg` option, make sure you pass in the Minikube profile name via `-mp` if the profile name is not `minikube`.

### Port forward

Before the setup, we will need to do a port-forward of the services that kiali is using.

We can use the `hack/run-kiali.sh` script for this purpose. To run it following the above schema:

```shell
./run-kiali.sh -pg 13000:3000 -pp 19090:9090 -app 8080 -es false -iu http://127.0.0.1:15014
```

> 

Reviewer: Just get rid of the extra line (yes, this is my OCD being triggered :)

```suggestion
```

### Local Configuration File

> The go process will require a configuration to point to these services and other specific configurations. 
> This file will be places in ~/kiali/config.yaml, and referenced later by the GO local process. 

Reviewer: fix typo, and mention the file in fixed font. Mention "Go" consistently (It is capitalized later on in this doc, so I just make it consistent with that here). Note that the file is placed in `/tmp` not `~` by default, see [here](https://github.com/kiali/kiali/blob/v1.51.0/hack/run-kiali.sh#L86)). You can change this location via the [option](https://github.com/kiali/kiali/blob/v1.51.0/hack/run-kiali.sh#L206-L208) `-trd` but that is not used in the command documented further up in this doc.

```suggestion
The Go process will require a configuration to point to these services and other specific configurations.
This file will be placed in `/tmp/kiali/config.yaml` by the `hack/run-kiali.sh` script, and will be referenced later by the Go local process. 
```

> ```yaml
> {{% readfile file="/static/files/contribution_guide/config.yaml" %}}
> ```

Reviewer: I am confused what this file is and why it is documented here. I don't think this is needed. Also, it has a setting that is very specific to a particular setup and not a general location that everyone will have - `static_content_root_directory: /home/userTests/kiali-static-files`. The hack/run-kiali.sh script will set this setting for you and will set it to a value that is valid for the user's environment. In addition, it sets up password auth for Jaeger tracing, but that is not enabled by default in most demo Istio set ups.

Instead, what I think you should do is point out the config template that the run-kiali.sh script will use just as a reference in case the user wants to look at it or need to edit it for a particular test scenario. Here's what I think this should be:

```suggestion
The `hack/run-kiali.sh` will generate the configuration file based on [this template file](https://github.com/kiali/kiali/blob/master/hack/run-kiali-config-template.yaml). You are free to modify this template for your own purposes.
```

### Local Processes 

In this section we will start the 3 local processes for kiali:
- kiali-core: The backend Go process
- kiali-ui: The frontend React process
- browser: The Javascript debugger process. 

We will need to fork the 3 kiali repositories, and then, clone them in a local folder:
- [Kiali](https://github.com/kiali/kiali)
- [Kiali-operator](https://github.com/kiali/kiali-operator)
- [Helm-charts](https://github.com/kiali/helm-charts)

In this example, we will create the configurations in the Jetbrains Golang IDE.

#### kiali-core
![kiali-core](/images/documentation/contribution/kiali-core.png)

#### kiali-ui

> In order to forward the requests to the backend propertly, we will need to add the following line in kiali/frontend/package.json:

Reviewer: fix typo:

```suggestion
In order to forward the requests to the backend properly, we will need to add the following line in kiali/frontend/package.json:
```

```json
"proxy": "http://localhost:8000",
```

![kiali-ui](/images/documentation/contribution/kiali-ui.png)

#### browser
![browser](/images/documentation/contribution/browser.png)

> After running the 3 processes, we should be able to access Kiali GUI in localhost:3000

Reviewer: Provide the URL - makes it more specific (it tells the user what URL scheme (http) to use and provides a link the user can click directly after following the steps).

```suggestion
After running the 3 processes, we should be able to access Kiali GUI at http://localhost:3000
```
